scipy/spatial/_lloyd_algorithm.py

```diff
@@ -0,0 +1,234 @@
+    qhull_options : str, optional
+        Additional options to pass to Qhull. See Qhull manual
+        for details. (Default: "Qbb Qc Qz Qj Qx" for ndim > 4 and
+        "Qbb Qc Qz Qj" otherwise.)
+
+    Returns
+    -------
+    points : array_like (n, d)
+        The points after being processed by Lloyd-Max algorithm.
+
+    Notes
+    -----
+    Lloyd-Max algorithm is an iterative process with the purpose of improving
+    the dispersion of points. For given points: (i) compute a Voronoi
+    Tessellation; (ii) find the centroid of each Voronoi cell; (iii) move the
+    points toward the centroid of their respective cell. See [1]_, [2]_.
+
+    The process converges to equally spaced points. It implies that measures
+    like the discrepancy could suffer from too many iterations. On the other
+    hand, L1 and L2 distances should improve. This is especially true with
+    QMC methods which tend to favor the discrepancy over other criteria.
+
+    .. note::
+
+        The current implementation does not intersect the Voronoi Tessellation
+        with the boundaries. This implies that for a low number of points,
+        empirically below 20, no Voronoi cell is touching the boundaries.
+        Hence points cannot be moved close to the boundaries.
+
+        Further improvements could consider the points at infinity so that
+        all boundaries are segments of some Voronoi cells. This would fix
+        the computation of the centroid position.
+
+    .. warning::
+
+       The Voronoi Tessellation step is expensive and quickly becomes
+       intractable with dimensions as low as 10 even for a sample of points
+       of size as low as 1000.
+
+    .. versionadded:: 1.8.0
+
+    References
+    ----------
+    .. [1] Lloyd. "Least Squares Quantization in PCM".
+       IEEE Transactions on Information Theory, 1982.
+    .. [2] Max J. "Quantizing for minimum distortion".
+       IEEE Transactions on Information Theory, 1960.
+
+    Examples
+    --------
+    >>> from scipy.spatial.distance import cdist
+    >>> from scipy.spatial import lloyd_centroidal_voronoi_tessellation
+    >>> rng = np.random.default_rng()
+    >>> points = rng.random((128, 2))
+
+    Compute the quality of the points using the L1 criterion.
+
+    >>> def l1_norm(points):
+    ...    l1 = cdist(points, points, 'cityblock')
+    ...    return np.min(l1[l1.nonzero()])
+
+    >>> l1_norm(points)
+    0.00161...  # random
+
+    Now process the points using Lloyd's algorithm and check the improvement
+    on the L1. The value should increase.
+
+    >>> points = lloyd_centroidal_voronoi_tessellation(points)
+    >>> l1_norm(points)
+    0.0278...  # random
+
+    """
+    points = np.asarray(points).copy()
+
+    if not points.ndim == 2:
+        raise ValueError('Sample is not a 2D array')
+
+    # Checking that points are within the bounds
+    if not ((points.max() <= 1.) and (points.min() >= 0.)):
```

Comment:
Wouldn't it be a better user experience if we handled an arbitrary input domain, rather than forcing users to normalize their distribution before processing?

Reply:
If we would like that we would need to introduce some parameter to provide the boundaries of the domain. I think it's better to tell the user to scale the domain. To help we have https://scipy.github.io/devdocs/reference/generated/scipy.stats.qmc.scale.html#scipy.stats.qmc.scale we could provide a link to this in the doc.

Reply:
Adding this link - or better yet, an example - would be a good compromise. Options like this can always be added.

Reply:
I will add a note with a link to the function for now.